lib/fluent/plugin/out_s3.rb

```diff
@@ -68,27 +87,51 @@ def configure(conf)
           Time.now.utc.strftime(path)
         }
       end
+
+      @storage_class = "REDUCED_REDUNDANCY" if @reduced_redundancy
     end
 
     def start
       super
       options = {}
-      if @aws_key_id && @aws_sec_key
+      credentials_options = {}
+      case
+      when @aws_key_id && @aws_sec_key
         options[:access_key_id] = @aws_key_id
         options[:secret_access_key] = @aws_sec_key
-      elsif ENV.key? "AWS_ACCESS_KEY_ID"
-        options[:credential_provider] = AWS::Core::CredentialProviders::ENVProvider.new('AWS')
+      when @assume_role_credentials
+        c = @assume_role_credentials
+        credentials_options[:role_arn] = c.role_arn
+        credentials_options[:role_session_name] = c.role_session_name
+        credentials_options[:policy] = c.policy if c.policy
+        credentials_options[:duration_seconds] = c.duration_seconds if c.duration_seconds
+        credentials_options[:external_id] = c.external_id if c.external_id
+        options[:credentials] = Aws::AssumeRoleCredentials.new(credentials_options)
+      when @instance_profile_credentials
```

Comment:
How about using same credentials if user specifies "aws_iam_retries" for backward compatibility?